### PR TITLE
feat(angular): Support Schematics

### PR DESCRIPTION
**Pull request by puku0x**
According to https://github.com/OnsenUI/OnsenUI/issues/2566 , this PR introduces Angular Schematics.

```
$ ng add ngx-onsenui
```

- installs `onsenui` and `ngx-onsenui`,
- adds OnsenModule and `CUSTOM_ELEMENTS_SCHEMA` in the root module,
- adds `onsenui.css` and `onsen-css-components.css` into angular.json.
- adds `ons-page` component selector in `src/tslint.json`

~~We can also generate an `ons-page` component using the schematics.~~ (Future work)

```
$ ng config cli.defaultCollection ngx-onsenui
$ ng generate page sample
```

### CODE REVIEW

**Comment by puku0x** [comment]
Aww, CI said some permission error was occurred and `rsync` was not found.
```
sh: 1: rsync: not found
```

**Comment by emccorson** [comment]
Sorry for the delay in replying. To be honest, I saw that Circle CI had
failed but then I forgot to reply.

It turns out the problem is rsync is not installed on the CI server. It
runs Debian 9.6 which I guess doesn't install rsync by default.

I'm not sure what the best course of action is here. If rsync is a program
you would expect most users to have installed then it should be fine to
just install rsync on the CI server. I think it probably is quite a common
program because I have it installed by default on my computer (Ubuntu
18.04).

What do you think?

2018年12月6日(木) 8:34 Noriyuki Shinpuku <notifications@github.com>:

**Comment by puku0x** [comment]
@emccorson Thank you for providing me the detail. I used `rsync` for copying the outputs of `ngx-onsenui` schematics because it is installed on my macOS too. I think it is well-known command to copy files and installing `rsync` is a feasible solution for the problem.

**Comment by emccorson** [comment]
@puku0x I've finally gotten round to reviewing this - sorry it has taken so long!

I'm having a bit of trouble running this locally.

I was trying to run `schematics .:page` from `bindings/angular2` directory but I got this error:

```
Error: Package "." was found but does not support schematics.
```

I think that was because there was no `schematics` key in `package.json` (it only exists in `package-dist.json`) so I added it and now I'm getting this error:

```
An error occured:
Error: Collection "." cannot be resolved.
    at NodeModulesEngineHost._resolveCollectionPath (/home/eric/.nvm/versions/node/v10.14.1/lib/node_modules/@angular-devkit/schematics-cli/node_modules/@angular-devkit/schematics/tools/node-module-engine-host.js:91:15)
    at NodeModulesEngineHost.createCollectionDescription (/home/eric/.nvm/versions/node/v10.14.1/lib/node_modules/@angular-devkit/schematics-cli/node_modules/@angular-devkit/schematics/tools/file-system-engine-host-base.js:111:27)
    at SchematicEngine._createCollectionDescription (/home/eric/.nvm/versions/node/v10.14.1/lib/node_modules/@angular-devkit/schematics-cli/node_modules/@angular-devkit/schematics/src/engine/engine.js:147:40)
    at SchematicEngine.createCollection (/home/eric/.nvm/versions/node/v10.14.1/lib/node_modules/@angular-devkit/schematics-cli/node_modules/@angular-devkit/schematics/src/engine/engine.js:140:43)
    at NodeWorkflow.execute (/home/eric/.nvm/versions/node/v10.14.1/lib/node_modules/@angular-devkit/schematics-cli/node_modules/@angular-devkit/schematics/src/workflow/base.js:95:41)
    at main (/home/eric/.nvm/versions/node/v10.14.1/lib/node_modules/@angular-devkit/schematics-cli/bin/schematics.js:162:24)
    at Object.<anonymous> (/home/eric/.nvm/versions/node/v10.14.1/lib/node_modules/@angular-devkit/schematics-cli/bin/schematics.js:253:5)
    at Module._compile (internal/modules/cjs/loader.js:688:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:699:10)
    at Module.load (internal/modules/cjs/loader.js:598:32)
```

Do you know any way I can fix this error? It's probably just something stupid I'm doing - I'm not very familiar with schematics.

Once I have it working from `bindings/angular2` with the `schematics` command, I'll try it out from another project with the `ng` command.

Other than that, the code looks good.

**Comment by puku0x** [comment]
I'm sorry for late reply.
I will fix this PR after updating the supported version of Angular to v8!

**Comment by puku0x** [comment]
I removed `ngx-onsenui:page` schematic to focus on `ng-add`.

**Comment by puku0x** [comment]
Finally, CI passed.

**Comment by puku0x** [comment]
I used `npm link` for local testing.
```
$ cd bindings/angular2
$ npm run build-dist
$ cd ..
$ npm link
```
```
$ ng n my-sandbox --routing=false --style=css
$ cd my-sandbox
$ ng add ngx-onsenui:ng-add
```

**Comment by puku0x** [comment]
@emccorson Could you review this PR again, please?

**Comment by emccorson** [comment]
> @emccorson Could you review this PR again, please?

Thanks for the PR update. Yes, I will do review it this week.

**Comment by emccorson** [comment]
@puku0x
I'm having a bit of trouble with these instructions to get it working locally:

1. From `bindings/angular2`, I ran `npm run build-dist` which creates the `dist` directory with the build of `ngx-onsenui`.
2. If I run `cd ..` then I would be in `bindings`, but that doesn't seem right because if I run `npm link` from that directory it links `onsenui`.
3. So I tried running `npm link` from `bindings/angular2/dist` which seems like the correct move because we have just generated `dist`.
4. Then I run `npm link ngx-onsenui` in the test sandbox.
5. But then `ng add ngx-onsenui:ng-add` gives me this error: 

```
Invalid tag name "ngx-onsenui:ng-add": Tags may not have any characters that encodeURIComponent encodes.
```

6. And `ng add ngx-onsenui` gives me this message:

```
Skipping installation: Package already installed
```

and then errors with:

```
Cannot find module '/home/eric/src/my-sandbox/node_modules/ngx-onsenui/schematics/ng-add/setup-project'
```

---

Sorry if I'm being stupid here, but do you know what the problem might be?

Also, I might be wrong, but I think for `ng add`, the syntax is normally

```
ng add <package-name>
```

not:

```
ng add <package-name>:ng-add
```

---

Again, I'm not that familiar with schematics so I might be totally wrong, but I would really appreciate whatever feedback you have. I'm looking forward to merging this PR!

**Comment by puku0x** [comment]
@emccorson 
I'm very sorry for the inconvenience. 🙇‍♂️
I misunderstood about `ng-add` and told you wrong information. 

You are right. `npm link` should be run in `bindings/angular2/dist` directory.
```
$ cd bindings/angular2
$ npm run build-dist
$ cd dist
$ npm link
```

Then, you can make a sandbox and link to it.
```
$ cd bindings/angular2
$ ng n my-app --routing --style=css
$ npm link ngx-onsenui
$ ng add ngx-onsenui
```

Please note `ngx-onsenui` should be linked again after running `ng-add` because the code in `ng-add` schematic which adds the required packages  ([ng-add/index.ts#L14-L15](https://github.com/puku0x/OnsenUI/blob/5df4cc55993526134f91d302d8d2fef201f6d09a/bindings/angular2/schematics/src/ng-add/index.ts#L14-L15)) installs `ngx-onsenui` from not local but NPM repository, and the link is reset. 

Therefore, I used `ng g` which is equivalent to [this line](https://github.com/puku0x/OnsenUI/blob/5df4cc55993526134f91d302d8d2fef201f6d09a/bindings/angular2/schematics/src/ng-add/index.ts#L18) after re-linking `ngx-onsenui` for testing the behavior after `ng-add`.
```
$ npm link ngx-onsenui
$ ng g ngx-onsenui:ng-add-setup-project
    ✅️ OnsenModule is imported
    ✅️ Added Onsen UI theme to styles
    🎉 Hooray! ngx-onsenui is successfully installed.
    ♨️ For more information, see https://onsen.io/v2/api/angular2/
```

File: .circleci/config.yml [hunk 1]
```diff
@@ -19,6 +19,9 @@ jobs:
       - run:
           name: Install coveralls and gulp globally
           command: sudo yarn global add coveralls gulp
+      - run:
+          name: Install rsync
+          command: sudo apt install -y rsync
```
Comment (emccorson) [inline .circleci/config.yml]: I don't think we have another instance of using `apt` to install dependencies in the Circle CI config, but this should be OK. If it breaks, we'll fix it at the time.

**Comment by puku0x** [comment]
@emccorson I see. Thank you very much!

**Pull request merged**